pkg/expr/nodes.go

```diff
@@ -250,47 +258,62 @@ func (dn *DSNode) Execute(ctx context.Context, now time.Time, _ mathexp.Vars, s
 		}
 	}
 
+	dataFrames, err := getResponseFrame(resp, dn.refID)
+	if err != nil {
+		return mathexp.Results{}, QueryError{
+			RefID:         dn.refID,
+			DatasourceUID: dn.datasource.UID,
+			Err:           err,
+		}
+	}
+
 	var result mathexp.Results
-	responseType, result, err = queryDataResponseToResults(ctx, resp, dn.refID, dn.datasource.Type, dn.datasource.UID, s)
+	responseType, result, err = convertDataFramesToResults(ctx, dataFrames, dn.datasource.Type, s, logger)
 	return result, err
 }
 
-func queryDataResponseToResults(ctx context.Context, resp *backend.QueryDataResponse, refID string, datasourceType string, datasourceUid string, s *Service) (string, mathexp.Results, error) {
-	vals := make([]mathexp.Value, 0)
+func getResponseFrame(resp *backend.QueryDataResponse, refID string) (data.Frames, error) {
 	response, ok := resp.Responses[refID]
 	if !ok {
-		if len(resp.Responses) > 0 {
-			keys := make([]string, 0, len(resp.Responses))
-			for refID := range resp.Responses {
-				keys = append(keys, refID)
-			}
-			logger.Warn("Can't find response by refID. Return nodata", "responseRefIds", keys)
+		// This indicates that the RefID of the request was not included to the response, i.e. some problem in the data source plugin
+		keys := make([]string, 0, len(resp.Responses))
+		for refID := range resp.Responses {
+			keys = append(keys, refID)
+		}
+		var availableRefIds string
+		if len(keys) > 0 {
+			sort.Strings(keys)
+			availableRefIds = fmt.Sprintf(", available refIDs [%s]", strings.Join(keys, ","))
 		}
-		return "no-data", mathexp.Results{Values: mathexp.Values{mathexp.NoData{}.New()}}, nil
+		return nil, fmt.Errorf("cannot find reference ID %s in response%s", refID, availableRefIds)
 	}
 
 	if response.Error != nil {
-		return "", mathexp.Results{}, QueryError{RefID: refID, Err: response.Error, DatasourceUID: datasourceUid}
+		return nil, response.Error
 	}
+	return response.Frames, nil
+}
 
+func convertDataFramesToResults(ctx context.Context, frames data.Frames, datasourceType string, s *Service, logger log.Logger) (string, mathexp.Results, error) {
+	vals := make([]mathexp.Value, 0)
```

Comment:
It seems okay looking at the code since we `range`, but maybe for the sake of simplicity we should return vals if `len(frames) == 0` at start of func

Reply:
Done